Platform/CommonBoardPkg/Library/BoardSupportLib/BoardSupportLib.c

```diff
@@ -0,0 +1,87 @@
+/** @file
+
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/PcdLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/ConfigDataLib.h>
+#include <Library/BoardSupportLib.h>
+#include <ConfigDataCommonDefs.h>
+
+#define  IMAGE_TYPE_ADDENDUM  0xFE
+#define  IMAGE_TYPE_NOT_USED  0xFF
+
+/**
+  Fill the boot option list data with CFGDATA info
+
+  @param[in, out]   OsBootOptionList pointer to boot option list.
+ **/
+VOID
+EFIAPI
+FillBootOptionListFromCfgData (
+  IN OUT   OS_BOOT_OPTION_LIST   *OsBootOptionList
+)
+{
+  OS_BOOT_OPTION             *BootOption;
+  OS_BOOT_OPTION             *BootOptionCfgData;
+  UINT8                       UpdateFlag;
+  UINTN                       ImageIdx;
+  UINTN                       Idx;
+  UINT64                      Lba;
+  CHAR8                      *StrPtr;
+
+  ImageIdx        = 1;
+  BootOption      = &OsBootOptionList->OsBootOption[0];
+  OsBootOptionList->CurrentBoot = 0;
+  for (Idx = 0; Idx < MAX_BOOT_OPTION_CFGDATA_ENTRY; Idx++) {
+    BootOptionCfgData = (OS_BOOT_OPTION *)FindConfigDataByTag (CDATA_BOOT_OPTION_TAG + Idx);
+    if (BootOptionCfgData == NULL) {
+      break;
+    }
+
+    if (BootOptionCfgData->ImageType == IMAGE_TYPE_NOT_USED) {
+      continue;
+    }
+
+    UpdateFlag = 0;
+    if (BootOptionCfgData->ImageType == IMAGE_TYPE_ADDENDUM) {
+      // This entry is an addendum for previous boot option entry
+      if (ImageIdx < ARRAY_SIZE(BootOption->Image)) {
+        UpdateFlag = 2;
+      }
+    } else {
+      // CFGDATA has short structure to save size on flash
+      // Need to translate the short format to OS_BOOT_OPTION format
+      UpdateFlag = 1;
+      ImageIdx   = 0;
+      CopyMem (BootOption, BootOptionCfgData, OFFSET_OF (OS_BOOT_OPTION, Image[0]));
+    }
+
+    if (UpdateFlag > 0) {
+      StrPtr = (CHAR8 *)BootOptionCfgData->Image[0].FileName;
+      // Use either LBA or filename. '#' indicates it is LBA string.
+      if ((StrPtr[0] == '#') && (AsciiStrHexToUint64S (StrPtr + 1, NULL, &Lba) == RETURN_SUCCESS)) {
+        BootOption->Image[ImageIdx].LbaImage.LbaAddr = Lba;
```

Comment:
Looks in the "addendum" boot option also has common header (e.g. devtype, hardware partition, etc). 
1) Do we want to have it or just have the information need in Image[0] to save size further?
2) Here it only copies LBA or filename here. where we copy other information, e.g. software partition?

Reply:
Yes, it has common header,  but not everything is used, such as devtype and hardware partition.  Since it is same structure as a normal entry, it has the same size.  Plus, the swpart will be used.  

Good catch, the swpart assignment was missing.  Will add it.

Reply:
Updated a new patch.

Reply:
Added logic in template to hide some items if it is configured as addendum.